include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h

```diff
@@ -0,0 +1,165 @@
+/*  ______________________________________________________________________
+ *
+ *  ExaDG - High-Order Discontinuous Galerkin for the Exa-Scale
+ *
+ *  Copyright (C) 2023 by the ExaDG authors
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+#ifndef EXADG_ACOUSTIC_CONSERVATION_EQUATIONS_SPATIAL_DISCRETIZATION_OPERATORS_WEAK_BOUNDARY_CONDITIONS_H_
+#define EXADG_ACOUSTIC_CONSERVATION_EQUATIONS_SPATIAL_DISCRETIZATION_OPERATORS_WEAK_BOUNDARY_CONDITIONS_H_
+
+#include <exadg/acoustic_conservation_equations/user_interface/boundary_descriptor.h>
+#include <exadg/functions_and_boundary_conditions/evaluate_functions.h>
+#include <exadg/matrix_free/integrators.h>
+
+namespace ExaDG
+{
+namespace Acoustics
+{
+/**
+ * Class to access boundary values with a similar interface to FaceIntegrators.
+ * This way, the operations on faces only has to be written once.
+ */
+template<int dim, typename Number>
+class BoundaryFaceIntegratorP
+{
+  using FaceIntegratorP = FaceIntegrator<dim, 1, Number>;
+
+public:
+  BoundaryFaceIntegratorP(FaceIntegratorP const &          integrator_m_in,
+                          BoundaryDescriptorP<dim> const & boundary_descriptor_in)
+    : integrator_m(integrator_m_in),
+      boundary_descriptor(boundary_descriptor_in),
+      evaluation_time(0.0),
+      boundary_id(dealii::numbers::invalid_boundary_id),
+      boundary_type(BoundaryTypeP::Undefined)
+  {
+  }
+
+  void
+  reinit(unsigned int const face)
+  {
+    auto const boundary_id_new = integrator_m.get_matrix_free().get_boundary_id(face);
+
+    // only update boundary_type if needed to avoid an unnecessary search in boundary_descriptor
+    if(boundary_id_new != boundary_id)
+    {
+      boundary_id   = boundary_id_new;
+      boundary_type = boundary_descriptor.get_boundary_type(boundary_id);
+    }
```

Comment:
This optimization could be shifted to `get_value()`. The (probably) expensive part `get_boundary_type()` (operating on `std::map`s) would not be done more often than currently, only the if-statement would be evaluated more often. Here, the question would again be whether such an if-statement has any relevance in terms of performance as compared to the other instructions/operations of `get_value()`.

The benefit would be that `BoundaryFaceIntegrator` does not depend on `BoundaryType` and `BoundaryDescriptor`, or that a base class `BoundaryFaceIntegratorBase` would not need `BoundaryType` and `BoundaryDescriptor` as template parameters.